### PR TITLE
Touch mirrors on even on fail to update

### PR DESCRIPTION
**Pull request by zeripath**
If a mirror fails to be synchronised it should be pushed to the bottom of the queue
of the awaiting mirrors to be synchronised. At present if there LIMIT number of
broken mirrors they can effectively prevent all other mirrors from being synchronized
as their last_updated time will remain earlier than other mirrors.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: models/repo/mirror.go [hunk 1]
```diff
@@ -113,6 +113,17 @@ func UpdateMirror(m *Mirror) error {
...
+func touchMirror(e db.Engine, m *Mirror) error {
+	m.UpdatedUnix = timeutil.TimeStampNow()
+	_, err := e.ID(m.ID).Cols("updated_unix").Update(m)
+	return err
+}
+
+// TouchMirror updates the mirror updatedUnix
+func TouchMirror(m *Mirror) error {
+	return touchMirror(db.GetEngine(db.DefaultContext), m)
+}
```
Comment (wxiaoguang) [inline models/repo/mirror.go]: Since we are going to use `ctx` everywhere, so I think a `TouchMirror(ctx context.Context, m *Mirror) error` is enough.

**Review by wxiaoguang — APPROVED** [review]

**Comment by lunny** [comment]
If there is only one mirror, what will happen?

File: models/repo/mirror.go (hunk 1 above)
Comment (zeripath) [reply models/repo/mirror.go]: The main reason I didn't do that is that I didn't want to refactor the rest of file, but I guess we can do that.

**Comment by zeripath** [comment]
> If there is only one mirror, what will happen?

It's fine - it will be tried again the next time round. The touch just causes the mirror to be reordered so that it is not the first mirror to be checked.

File: services/mirror/mirror_pull.go [hunk 2]
```diff
@@ -399,6 +399,10 @@ func SyncPullMirror(ctx context.Context, repoID int64) bool {
 	log.Trace("SyncMirrors [repo: %-v]: Running Sync", m.Repo)
 	results, ok := runSync(ctx, m)
 	if !ok {
+		if err = repo_model.TouchMirror(ctx, m); err != nil {
+			log.Error("SyncMirrors [repo: %-v]: failed to TouchMirror: %v", m.Repo, err)
+			return false
```
Comment (lunny) [inline services/mirror/mirror_pull.go]: This line could be removed.

Comment (zeripath) [reply services/mirror/mirror_pull.go]:
```suggestion
```

**Review by delvh — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Pull request merged**